storage: create a protobuf to hold stores

andrewbaptist opened the pull request:
Summary of changes:
* Create a struct configpb.Storage to represent all the "store related" pieces (mostly store config and WAL failover)
* Move the encryption code into the configpb directory and delay "serialization" of the encryption options. 
* Remove all direct reads of the `Stores` from `server.Config`, instead use the new `StorageConfig`
* Various cleanups to pass the `configpb.Storage` or `configpb.Store` in construction
* Persisting the StorageConfig on each of the stores
* Addition of a `bootstrapConfig` option which allows reading the config from either the OS disk or one of the existing disks.

WIP: This commit will be split up eventually.
Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/138051)
<!-- Reviewable:end -->

Commit:
storage: create a protobuf to hold stores

Summary of changes:

* Create a struct configpb.Storage to represent all the "store related" pieces (mostly store config and WAL failover)
* Move the encryption code into the configpb directory and delay "serialization" of the encryption options.
* Remove all direct reads of the Stores from server.Config, instead use the new StorageConfig
* Various cleanups to pass the configpb.Storage or configpb.Store in construction
* Persisting the StorageConfig on each of the stores
* Addition of a bootstrapConfig option which allows reading the config from either the OS disk or one of the existing disks.

WIP: This commit will be split up eventually.

Epic: none

Release note: None

Commit:
storage: remove storage-engine flag

The only storage engine for the past 4 years has been pebble. The
storage engine is specfied at a node level but applies to the stores. If
we want to add something like storage engine back in the future it would
be done differently so its easiest to simply remove this for now.

Epic: CRDB-41111

Release note (cli change): This change removes the ``--storage-engine``
parameter from the CLI. It has been deprecated for over 4 years.
